### PR TITLE
Latent regeneration attack + setup material

### PR DESCRIPTION
**Pull request by tomputer-g**
Added Regeneration/Rinsing attack scripts from the WAVES & WatermarkAttacker papers. requirements.txt should contain all dependencies. Conda environment file is there but something with sympy prevents the script from importing something from diffusers package, so we should use venv for now.

I test ran this on the example watermarked image with the setup from scratch so everything should work. Documentation on how to run the script is provided in the README. Let me know if I should add anything else.

### CODE REVIEW

File: latent_rinsing_attack/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# What?
+
+This is an adaptation of the WAVES Regeneration attack inspired by Rinse-4xdiff. We also cite the WatermarkAttacker repository, from which ReSDPipeline class is used.
```
Comment (rmdluo) [inline latent_rinsing_attack/README.md:3]: Maybe consider adding the citations/links to WAVES and WatermarkAttacker

File: latent_rinsing_attack/conda_env.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+name: latent_rinsing_env
+channels:
```
Comment (rmdluo) [inline latent_rinsing_attack/conda_env.yml]: when we merge this, I'll make an issue for this

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,6 @@
+# adversarial checkpoint
+adversarial_attack/surrogate_models/
```
Comment (rmdluo) [inline .gitignore]: maybe put the project specific files you want to ignore in nested .gitignore files in the other project directories

File: latent_rinsing_attack/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+The Conda install yaml doesn't currently work due to some import issue, so a venv is used for now.
+
+Run the attack:
+
+```bash
+# rinse.py watermarked_path output_path strength n_rinse [--model StableDiffusionMODEL]
+# Model can be "CompVis/stable-diffusion-v1-4" or (by default) "stabilityai/stable-diffusion-2-1" or something else
+# Output folder will be created if it doesn't already exist.
+python rinse.py ./watermarked_example/ ./attacked_example 10 1
+```
```
Comment (rmdluo) [inline latent_rinsing_attack/README.md]: can you add some stuff on how to do metricing?

**Review by rmdluo — COMMENTED** [review]
Code runs, need some more info in the readme. Also, maybe try the conda env

```
name: latent_rinsing_env
channels:
  - pytorch
  - nvidia
  - conda-forge
  - defaults
dependencies:
  - python=3.10
  - pip
  - tqdm
  - pytorch
  - pytorch-cuda=11.8
  - pip:
    - diffusers
    - transformers
    - accelerate
```

conda env create -f conda_env.yml (possibly rename to environment.yml for consistency)

File: latent_rinsing_attack/conda_env.yml (hunk 2 above)
Comment (rmdluo) [reply latent_rinsing_attack/conda_env.yml]: see conda env in overall review

File: latent_rinsing_attack/README.md (hunk 1 above)
Comment (tomputer-g) [reply latent_rinsing_attack/README.md:3]: Those were added in the script, but good point. I'll also add the links there

File: latent_rinsing_attack/conda_env.yml (hunk 2 above)
Comment (tomputer-g) [reply latent_rinsing_attack/conda_env.yml]: Thanks, will verify on my end too

File: .gitignore (hunk 3 above)
Comment (tomputer-g) [reply .gitignore]: Missed it when copying over from a sandbox folder. I'll clean this. Thanks

File: latent_rinsing_attack/README.md (hunk 4 above)
Comment (tomputer-g) [reply latent_rinsing_attack/README.md]: Metrics as in FID, etc.? Will ask offline on exactly what we need

File: .gitignore (hunk 3 above)
Comment (tomputer-g) [reply .gitignore]: This was due to merging the adversairla attack branch into mine. I'll remove everything adversarial-wise from this PR

File: latent_rinsing_attack/README.md (hunk 4 above)
Comment (tomputer-g) [reply latent_rinsing_attack/README.md]: Now warm-up-kit is a submodule in this repo, adding comment to README to specify how to use it for evaluation

**Pull request merged**